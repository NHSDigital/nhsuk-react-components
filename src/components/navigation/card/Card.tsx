import React, { FC, HTMLProps } from 'react';
import classNames from 'classnames';
import CardContext from './CardContext';
import CardContent from './components/CardContent';
import CardDescription from './components/CardDescription';
import CardImage from './components/CardImage';
import CardLink from './components/CardLink';
import CardHeading from './components/CardHeading';
import CardGroup from './components/CardGroup';
import CardGroupItem from './components/CardGroupItem';
import { CardType } from '@util/types/NHSUKTypes';
import { cardTypeIsCareCard } from '@util/types/TypeGuards';

interface CardProps extends HTMLProps<HTMLDivElement> {
  clickable?: boolean;
  cardType?: CardType;
}

interface ICard extends FC<CardProps> {
  Content: typeof CardContent;
  Description: typeof CardDescription;
  Image: typeof CardImage;
  Link: typeof CardLink;
  Heading: typeof CardHeading;
  Group: typeof CardGroup;
  GroupItem: typeof CardGroupItem;
}

const Card: ICard = ({ className, clickable, children, cardType, ...rest }) => {
  let cardClassNames = classNames(
    'nhsuk-card',
    { 'nhsuk-card--clickable': clickable },
    { 'nhsuk-card--feature': cardType === 'feature' },
    { 'nhsuk-card--secondary': cardType === 'secondary' },
    className,
  );

  if (cardTypeIsCareCard(cardType)) {
    cardClassNames = classNames(
      cardClassNames,
      'nhsuk-card--care',
      `nhsuk-card--care--${cardType}`,
    );
  }

  return (
    <div className={cardClassNames} {...rest}>
      <CardContext.Provider
        value={{
          cardType,
        }}
      >
        {children}
      </CardContext.Provider>
    </div>
  );
};

Card.Heading = CardHeading;
Card.Description = CardDescription;
Card.Image = CardImage;
Card.Link = CardLink;
Card.Content = CardContent;
Card.Group = CardGroup;
Card.GroupItem = CardGroupItem;

export default Card;
