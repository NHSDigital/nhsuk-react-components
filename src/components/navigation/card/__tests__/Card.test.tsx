/* eslint-disable jsx-a11y/anchor-is-valid */
import { render } from '@testing-library/react';
import React from 'react';
import Card from '../Card';

describe('Card', () => {
  it('matches snapshot', () => {
    const { container } = render(
      <Card>
        <Card.Image src="imageSrc" alt="imageAlt" />
        <Card.Content>
          <Card.Heading>If you need help now but it&apos;s not an emergency</Card.Heading>
          <Card.Description>
            Go to <a href="#">111.nhs.uk</a> or <a href="#">call 111</a>
          </Card.Description>
        </Card.Content>
      </Card>,
    );

    expect(container).toMatchSnapshot();
  });

  it('can render Card.Link as different elements', () => {
    const { container } = render(
      <Card clickable>
        <Card.Content>
          <Card.Heading>
            <Card.Link asElement="button" type="button">
              Click me!
            </Card.Link>
          </Card.Heading>
        </Card.Content>
      </Card>,
    );

    expect(container.querySelector('button.nhsuk-card__link')?.textContent).toBe('Click me!');
  });

  it('adds clickable classes', () => {
    const { container } = render(
      <Card clickable>
        <Card.Content>
          <Card.Heading className="nhsuk-heading-m">
            <Card.Link href="#">Introduction to care and support</Card.Link>
          </Card.Heading>
          <Card.Description>
            A quick guide for people who have care and support needs and their carers
          </Card.Description>
        </Card.Content>
      </Card>,
    );

    expect(container.querySelector('div.nhsuk-card.nhsuk-card.nhsuk-card--clickable')).toBeTruthy();
  });

  it('adds feature classes to all elements', () => {
    const { container } = render(
      <Card cardType="feature">
        <Card.Content>
          <Card.Heading>Feature card heading</Card.Heading>
          <Card.Description>Feature card description</Card.Description>
        </Card.Content>
      </Card>,
    );

    expect(container.querySelector('div.nhsuk-card.nhsuk-card.nhsuk-card--feature')).toBeTruthy();
    expect(
      container.querySelector(
        'div.nhsuk-card__content.nhsuk-card__content.nhsuk-card__content--feature',
      ),
    ).toBeTruthy();
    expect(
      container.querySelector(
        'h2.nhsuk-card__heading.nhsuk-card__heading.nhsuk-card__heading--feature',
      ),
    ).toBeTruthy();
  });

  it('adds primary class to card contents', () => {
    const { container } = render(
      <Card cardType="primary">
        <Card.Content>
          <Card.Heading>Feature card heading</Card.Heading>
          <Card.Description>Feature card description</Card.Description>
        </Card.Content>
      </Card>,
    );

    expect(
      container.querySelector(
        'div.nhsuk-card__content.nhsuk-card__content.nhsuk-card__content--primary',
      ),
    ).toBeTruthy();
  });

  it('adds secondary classes to card and contents', () => {
    const { container } = render(
      <Card cardType="secondary">
        <Card.Content>
          <Card.Heading>Feature card heading</Card.Heading>
          <Card.Description>Feature card description</Card.Description>
        </Card.Content>
      </Card>,
    );

    expect(container.querySelector('div.nhsuk-card.nhsuk-card.nhsuk-card--secondary')).toBeTruthy();
    expect(
      container.querySelector(
        'div.nhsuk-card__content.nhsuk-card__content.nhsuk-card__content--secondary',
      ),
    ).toBeTruthy();
  });

  describe('Care card variant', () => {
    describe('non-urgent', () => {
      it('matches the snapshot', () => {
        const { container } = render(
          <Card cardType="non-urgent">
            <Card.Heading>Non urgent heading</Card.Heading>
          </Card>,
        );

        expect(container).toMatchSnapshot();
      });
      it('adds classes to card', () => {
        const { container } = render(
          <Card cardType="non-urgent">
            <Card.Heading>Non urgent heading</Card.Heading>
          </Card>,
        );

        expect(container.querySelector('div.nhsuk-card.nhsuk-card.nhsuk-card--care')).toBeTruthy();
        expect(
          container.querySelector('div.nhsuk-card.nhsuk-card.nhsuk-card--care--non-urgent'),
        ).toBeTruthy();
      });

      it('renders the heading with the expected hidden text', () => {
        const { container } = render(
          <Card cardType="non-urgent">
            <Card.Heading>Non urgent heading</Card.Heading>
          </Card>,
        );

        const headingContainer = container.querySelector('.nhsuk-card--care__heading-container');

        expect(headingContainer?.querySelector('.nhsuk-u-visually-hidden')?.textContent).toEqual(
          'Non-urgent advice: ',
        );
      });
    });

    describe('urgent', () => {
      it('matches the snapshot', () => {
        const { container } = render(
          <Card cardType="urgent">
            <Card.Heading>Urgent heading</Card.Heading>
          </Card>,
        );

        expect(container).toMatchSnapshot();
      });
      it('adds classes to card', () => {
        const { container } = render(
          <Card cardType="urgent">
            <Card.Heading>Urgent heading</Card.Heading>
          </Card>,
        );

        expect(container.querySelector('div.nhsuk-card.nhsuk-card.nhsuk-card--care')).toBeTruthy();
        expect(
          container.querySelector('div.nhsuk-card.nhsuk-card.nhsuk-card--care--urgent'),
        ).toBeTruthy();
      });

      it('renders the heading with the expected hidden text', () => {
        const { container } = render(
          <Card cardType="urgent">
            <Card.Heading>Urgent heading</Card.Heading>
          </Card>,
        );

        const headingContainer = container.querySelector('.nhsuk-card--care__heading-container');

        expect(headingContainer?.querySelector('.nhsuk-u-visually-hidden')?.textContent).toEqual(
          'Urgent advice: ',
        );
      });
    });

    describe('emergency', () => {
      it('matches the snapshot', () => {
        const { container } = render(
          <Card cardType="emergency">
            <Card.Heading>Emergency heading</Card.Heading>
          </Card>,
        );

        expect(container).toMatchSnapshot();
      });
      it('adds classes to card', () => {
        const { container } = render(
          <Card cardType="emergency">
            <Card.Heading>Emergency heading</Card.Heading>
          </Card>,
        );

        expect(container.querySelector('div.nhsuk-card.nhsuk-card.nhsuk-card--care')).toBeTruthy();
        expect(
          container.querySelector('div.nhsuk-card.nhsuk-card.nhsuk-card--care--emergency'),
        ).toBeTruthy();
      });

      it('renders the heading with the expected hidden text', () => {
        const { container } = render(
          <Card cardType="emergency">
            <Card.Heading>Emergency heading</Card.Heading>
          </Card>,
        );

        const headingContainer = container.querySelector('.nhsuk-card--care__heading-container');

        expect(headingContainer?.querySelector('.nhsuk-u-visually-hidden')?.textContent).toEqual(
          'Immediate action required: ',
        );
      });
    });

    describe('hidden text', () => {
      it('renders without hidden text', () => {
        const { container } = render(
          <Card cardType="urgent">
            <Card.Heading visuallyHiddenText={false}>Heading</Card.Heading>
          </Card>,
        );

        expect(container.querySelector('.nhsuk-u-visually-hidden')).toBeFalsy();
      });

      it('renders with hidden text', () => {
        const { container } = render(
          <Card cardType="urgent">
            <Card.Heading visuallyHiddenText="Custom">Heading</Card.Heading>
          </Card>,
        );

        expect(container.querySelector('.nhsuk-u-visually-hidden')?.textContent).toEqual('Custom');
      });
    });
  });

  describe('Card.Group', () => {
    it('matches snapshot', () => {
      const { container } = render(
        <Card.Group>
          <Card.GroupItem width="one-half">
            <Card>
              <Card.Content>
                <Card.Heading>Test Card 1</Card.Heading>
                <Card.Description>Test Card 1 Description</Card.Description>
              </Card.Content>
            </Card>
          </Card.GroupItem>
          <Card.GroupItem width="one-half">
            <Card>
              <Card.Content>
                <Card.Heading>Test Card 2</Card.Heading>
                <Card.Description>Test Card 2 Description</Card.Description>
              </Card.Content>
            </Card>
          </Card.GroupItem>
        </Card.Group>,
      );

      expect(container).toMatchSnapshot();
    });
  });
});
