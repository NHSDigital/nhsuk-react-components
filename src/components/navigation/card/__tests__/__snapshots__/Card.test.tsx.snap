// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Card.Group matches snapshot 1`] = `
<div>
  <div
    class="nhsuk-grid-row nhsuk-card-group"
  >
    <div
      class="nhsuk-grid-column-one-half nhsuk-card-group__item"
    >
      <div
        class="nhsuk-card"
      >
        <div
          class="nhsuk-card__content"
        >
          <h2
            class="nhsuk-card__heading"
          >
            Test Card 1
          </h2>
          <p
            class="nhsuk-card__description"
          >
            Test Card 1 Description
          </p>
        </div>
      </div>
    </div>
    <div
      class="nhsuk-grid-column-one-half nhsuk-card-group__item"
    >
      <div
        class="nhsuk-card"
      >
        <div
          class="nhsuk-card__content"
        >
          <h2
            class="nhsuk-card__heading"
          >
            Test Card 2
          </h2>
          <p
            class="nhsuk-card__description"
          >
            Test Card 2 Description
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Card Care card variant emergency matches the snapshot 1`] = `
<div>
  <div
    class="nhsuk-card nhsuk-card--care nhsuk-card--care--emergency"
  >
    <div
      class="nhsuk-card--care__heading-container"
    >
      <h2
        class="nhsuk-card--care__heading"
      >
        <span
          role="text"
        >
          <span
            class="nhsuk-u-visually-hidden"
          >
            Immediate action required: 
          </span>
          Emergency heading
        </span>
      </h2>
      <span
        aria-hidden="true"
        class="nhsuk-card--care__arrow"
      />
    </div>
  </div>
</div>
`;

exports[`Card Care card variant non-urgent matches the snapshot 1`] = `
<div>
  <div
    class="nhsuk-card nhsuk-card--care nhsuk-card--care--non-urgent"
  >
    <div
      class="nhsuk-card--care__heading-container"
    >
      <h2
        class="nhsuk-card--care__heading"
      >
        <span
          role="text"
        >
          <span
            class="nhsuk-u-visually-hidden"
          >
            Non-urgent advice: 
          </span>
          Non urgent heading
        </span>
      </h2>
      <span
        aria-hidden="true"
        class="nhsuk-card--care__arrow"
      />
    </div>
  </div>
</div>
`;

exports[`Card Care card variant urgent matches the snapshot 1`] = `
<div>
  <div
    class="nhsuk-card nhsuk-card--care nhsuk-card--care--urgent"
  >
    <div
      class="nhsuk-card--care__heading-container"
    >
      <h2
        class="nhsuk-card--care__heading"
      >
        <span
          role="text"
        >
          <span
            class="nhsuk-u-visually-hidden"
          >
            Urgent advice: 
          </span>
          Urgent heading
        </span>
      </h2>
      <span
        aria-hidden="true"
        class="nhsuk-card--care__arrow"
      />
    </div>
  </div>
</div>
`;

exports[`Card matches snapshot 1`] = `
<div>
  <div
    class="nhsuk-card"
  >
    <img
      alt="imageAlt"
      class="nhsuk-card__img"
      src="imageSrc"
    />
    <div
      class="nhsuk-card__content"
    >
      <h2
        class="nhsuk-card__heading"
      >
        If you need help now but it's not an emergency
      </h2>
      <p
        class="nhsuk-card__description"
      >
        Go to 
        <a
          href="#"
        >
          111.nhs.uk
        </a>
         or 
        <a
          href="#"
        >
          call 111
        </a>
      </p>
    </div>
  </div>
</div>
`;
